server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingService.java

```diff
@@ -31,12 +31,61 @@
 import org.apache.james.mailrepository.api.MailRepositoryStore;
 import org.apache.james.queue.api.MailQueue;
 import org.apache.james.queue.api.MailQueueFactory;
+import org.apache.james.util.OptionalUtils;
 import org.apache.james.util.streams.Iterators;
 import org.apache.mailet.Mail;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.github.fge.lambdas.Throwing;
 
 public class ReprocessingService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReprocessingService.class);
+
+    public static class MissingKeyException extends RuntimeException {
+        MissingKeyException(MailKey key) {
+            super(key.asString() + " can not be found");
+        }
+    }
+
+    interface MissingMailPolicy {
+        Optional<Mail> validate(Optional<Mail> mail, MailKey key);
+    }
+
+    static class Performer {
+        private final MissingMailPolicy missingMailPolicy;
+        private final MailQueue mailQueue;
+        private final Optional<String> targetProcessor;
+
+        Performer(MissingMailPolicy missingMailPolicy, MailQueue mailQueue, Optional<String> targetProcessor) {
+            this.missingMailPolicy = missingMailPolicy;
+            this.mailQueue = mailQueue;
+            this.targetProcessor = targetProcessor;
+        }
+
+        private void reprocess(MailRepository repository, MailKey key) {
```

Comment:
reprocess could return a type T defined by the MissingMailPolicy so that we can implement a ValidationMissingMailPolicy that actually returns a Validation object (which aggregates errors).
WDYT?

Reply:
A missing key is not really an error.... CF the test that I implemented for the reprocessing service.